### ISSUE TITLE
New Rule: usage of Promise as final result (non-await'ed)

### ISSUE THREAD

**Issue by rentalhost** [eslint/eslint#16668]
### Rule details

This should verify that an `async function` return has been used without `await` on final occasions, such as `if()`.

### Related ECMAScript feature

`await`, `async`, `Promise`

### What type of rule is this?

Warns about a potential problem

### Example code

```js
const isValid = async () => {
    return new Promise((resolve) => {
        setTimeout(() => resolve(false), 50)
    })
}

const testIncorrect = async () => {
    if (isValid()) {
        return console.log('testIncorrect: Ops! Actually it is not valid!')
            // ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ INCORRECT!
    }

    console.log('testIncorrect: Expected')
}

const testCorrect = async () => {
    if (await isValid()) {
        return console.log('testCorrect: Ops! Actually it is not valid!')
    }

    console.log('testCorrect: Expected')
 // ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ CORRECT!
}

(async () => {
	await testIncorrect()
	await testCorrect()
})()

// testIncorrect: Ops! Actually it is not valid!
// testCorrect: Expected
```

### Why should this rule be in the core instead of a plugin?

It is directly related to EcmaScript features: `await`, `async` and `Promise`, and because it is a common error and difficult to analyze, it would be great to have this native in eslint.

### Participation

- [ ] I am willing to submit a pull request to implement this rule.

**Comment by bmish** [comment]
Does [require-await](https://eslint.org/docs/latest/rules/require-await) help?

**Comment by rentalhost** [comment]
@bmish Actually, no.

The `require-await` rule checks if an `async function` contains an `await` keyword in fact. On my example, I need `async function` but not _await_ for anything else inside this `function`, but returns a `Promise` in fact that must be `await`'ed for an external `function` (caller).

* `require-await` => `caller() => async () => await x()` (_correct_)
* `require-await` => `caller() => async () => x()` (_incorrect_)
* `new-rule` => `caller() => if(await (async () => x()))` (_correct_)
* `new-rule` => `caller() => if(async () => x())` (_incorrect_)

**Comment by mdjermanovic** [comment]
> This should verify that an `async function` return has been used without `await` on final occasions

I think this is in the realm of type checking, rather than linting. In fact, typescript can catch these errors:

[https://www.typescriptlang.org/play?#code/...](https://www.typescriptlang.org/play?#code/MYewdgzgLgBAlhAagQwDZwCYwLw2RATzGBgAoBKHAPhgG8AoGJmAJwFMoBXFsGMNgO4wACixABbBG1Kl2EEKgBubSthoNmmmBA4AVOOLYhOUGappyFy0gDM0O8gBoYAVgAM5RswC+n7-XpQSFgoNmgASWIQFnZgWFx8IhIKajovJjgbMgQUdAwKSg0tJnYuHhgg+VQ2ADpUEABzUgByUIiomLY4gC4YAHkABwgAQhgAQTjONFQCeFgEPhBYRTRMYebPYuKAem2YAD1Do+OT07Pzi8uTmHCAOQBhPoAlJ4BRe91h9Jh-b8qFWr1JqtMJQSKgTo9GCvAAeAy6oQwG3o3iAA)

```
This condition will always return true since this 'Promise<unknown>' is always defined.
```

**Comment by rentalhost** [comment]
@mdjermanovic You are right, but it fails with negative conditions. It must not be an error, infact, but is not common you check if for `false` when a `Promise` is returned:

https://www.typescriptlang.org/play?#code/MYewdgzgLgBAlhAkmAbgQwDZwCYwLwxoQCeYwMAFAJT4B8MA3gFAyswBOAplAK7tgwwnAO4wACuxABbBJwoUuEEBhScaeeszbaYEbgBU4UziB5R56+ouWqKAM0x6qAGhgBWAAxUWbAL7ffJiZQSFgoTmhkUHYuYFgCIlJyajpGH1Y4O0oAQgRkdCxsahoGGAB6MpgAORAYThiQdmyYAC4Acgp0mEDfIA

**Comment by JoshuaKGoldberg** [comment]
Is this https://typescript-eslint.io/rules/return-await?

**Comment by rentalhost** [comment]
@JoshuaKGoldberg no. Same as before.

**Comment by rentalhost** [comment]
I think this can be finished. Using `@typescript-eslint` the problem is detected.

![image](https://user-images.githubusercontent.com/421851/208253174-aace1c0e-1aa2-4b0a-94bd-034f7a940973.png)
